osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/datadecoder/BasicDlmsDataDecoder.java

```diff
@@ -0,0 +1,166 @@
+    final Long value = this.dlmsHelper.readLong(attributeData, "read number");
+
+    if (value == null) {
+      return "null";
+    } else {
+      return value.toString();
+    }
+  }
+
+  private String decodeBoolean(final DataObject attributeData) {
+    return attributeData.getValue().toString();
+  }
+
+  private String decodeOctetString(final DataObject attributeData) {
+
+    try {
+      return this.decodeDateTime(attributeData);
+    } catch (final Exception ignored) {
+      // Decoding failed, so byte array is not a date time
+    }
+
+    final byte[] byteArray = attributeData.getValue();
+    return new String(byteArray);
+  }
+
+  public String decodeDateTime(final DataObject attributeData) throws ProtocolAdapterException {
+    return this.cosemDateTimeToString(
+        this.dlmsHelper.readDateTime(attributeData, "Try to read time from bytes"));
+  }
+
+  public String decodeDate(final DataObject attributeData) {
+    if (attributeData.getType() == Type.OCTET_STRING) {
+
+      final ByteBuffer bb = ByteBuffer.wrap(attributeData.getValue());
+
+      final int year = bb.getShort() & 0xFFFF;
+      final int month = bb.get() & 0xFF;
+      final int dayOfMonth = bb.get() & 0xFF;
+      final int dayOfWeek = bb.get() & 0xFF;
+
+      return this.getDayOfWeek(dayOfWeek)
+          + ", "
+          + ((year == 65535) ? "Year not specified" : year)
+          + "-"
+          + month
+          + "-"
+          + dayOfMonth;
+    }
+    return "not a valid date";
+  }
+
+  public String decodeTime(final DataObject attributeData) {
+    if (attributeData.getType() == Type.OCTET_STRING) {
+
+      final ByteBuffer bb = ByteBuffer.wrap(attributeData.getValue());
+
+      final int hour = bb.get() & 0xFF;
+      final int minute = bb.get() & 0xFF;
+      final int second = bb.get() & 0xFF;
+      final int hundredths = bb.get() & 0xFF;
+
+      return String.format("%d:%02d:%02d.%03d", hour, minute, second, hundredths);
+    }
+    return "not a valid time";
+  }
+
+  private String cosemDateTimeToString(final CosemDateTimeDto dateTimeDto) {
+    return this.getDayOfWeek(dateTimeDto)
+        + dateTimeDto.getDate().asLocalDate()
+        + ", "
+        + dateTimeDto.getTime().asLocalTime().toString()
+        + ", deviation="
+        + dateTimeDto.getDeviation()
+        + ", "
+        + dateTimeDto.getClockStatus()
+        + ']';
+  }
+
+  private String getDayOfWeek(final CosemDateTimeDto dateTimeDto) {
+    return this.getDayOfWeek(dateTimeDto.getDate().getDayOfWeek());
+  }
+
+  private String getDayOfWeek(final int dayOfWeek) {
+    return switch (dayOfWeek) {
+      case 1 -> "Monday";
+      case 2 -> "Tuesday";
+      case 3 -> "Wednesday";
+      case 4 -> "Thursday";
+      case 5 -> "Friday";
+      case 6 -> "Saturday";
+      case 7 -> "Sunday";
+      case 0xFF -> "Day of week not specified";
+      default -> "DayOfWeek value unknown: " + dayOfWeek + ", ";
```

Comment:
Enumeration for commonly used translation?

Reply:
Created an Enum in a separate public class.